pkg/kv/kvserver/loqrecovery/server.go

```diff
@@ -0,0 +1,205 @@
+	// known cluster node.
+	// If visitor returns error marked with errMarkRetry then node would be dialed
+	// again and revisited until visitor succeeds or retry attempts are exhausted.
+	// In latter case cause of retryVisit is returned.
+	// Returns error if visitor failed or if error happened when processing node
+	// data.
+	VisitAvailableNodesWithRetry(ctx context.Context, retryOpts retry.Options,
+		visitor func(nodeID roachpb.NodeID, client serverpb.AdminClient) error,
+	) error
+}
+
+type Server struct {
+	stores               *kvserver.Stores
+	metadataQueryTimeout time.Duration
+	forwardReplicaFilter func(*serverpb.RecoveryCollectLocalReplicaInfoResponse) error
+}
+
+func NewServer(stores *kvserver.Stores, knobs base.ModuleTestingKnobs) *Server {
+	// Server side timeouts are necessary in recovery collector since we do best
+	// effort operations where cluster info collection as an operation succeeds
+	// even if some parts of it time out.
+	metadataQueryTimeout := 1 * time.Minute
+	var forwardReplicaFilter func(*serverpb.RecoveryCollectLocalReplicaInfoResponse) error
+	if rk, ok := knobs.(*TestingKnobs); ok {
+		if rk.MetadataScanTimeout > 0 {
+			metadataQueryTimeout = rk.MetadataScanTimeout
+		}
+		forwardReplicaFilter = rk.ForwardReplicaFilter
+	}
+	return &Server{
+		stores:               stores,
+		metadataQueryTimeout: metadataQueryTimeout,
+		forwardReplicaFilter: forwardReplicaFilter,
+	}
+}
+
+func (s Server) ServeLocalReplicas(
+	ctx context.Context,
+	_ *serverpb.RecoveryCollectLocalReplicaInfoRequest,
+	stream serverpb.Admin_RecoveryCollectLocalReplicaInfoServer,
+) error {
+	return s.stores.VisitStores(func(s *kvserver.Store) error {
+		reader := s.Engine().NewSnapshot()
+		defer reader.Close()
+		return visitStoreReplicas(ctx, reader, s.StoreID(), s.NodeID(),
+			func(info loqrecoverypb.ReplicaInfo) error {
+				return stream.Send(&serverpb.RecoveryCollectLocalReplicaInfoResponse{ReplicaInfo: &info})
+			})
+	})
+}
+
+func (s Server) ServeClusterReplicas(
+	ctx context.Context,
+	_ *serverpb.RecoveryCollectReplicaInfoRequest,
+	outStream serverpb.Admin_RecoveryCollectReplicaInfoServer,
+	kvDB *kv.DB,
+	clusterAdminClient ClusterAdminClient,
+) (err error) {
+	var (
+		descriptors, nodes, replicas int
+	)
+	defer func() {
+		if err == nil {
+			log.Infof(ctx, "streamed info: range descriptors %d, nodes %d, replica infos %d", descriptors,
+				nodes, replicas)
+		}
+	}()
+
+	err = contextutil.RunWithTimeout(ctx, "scan-range-descriptors", s.metadataQueryTimeout,
+		func(txnCtx context.Context) error {
+			txn := kvDB.NewTxn(txnCtx, "scan-range-descriptors")
+			if err := txn.SetFixedTimestamp(txnCtx, kvDB.Clock().Now()); err != nil {
+				return err
+			}
+			defer func() { _ = txn.Rollback(txnCtx) }()
+			log.Infof(txnCtx, "serving recovery range descriptors for all ranges")
+			return txn.Iterate(txnCtx, keys.Meta2Prefix, keys.MetaMax, rangeMetadataScanChunkSize,
```

Comment:
How does this interact with retryable transaction errors? Are they automatically retried, and if so, will `txn.Iterate()` emit duplicates? If not, will this be flaky in the face of retryable errors, since we don't handle them?

Reply:
This should also be handled by setting timestamp. We should only be getting non retryable errors here and we are fine then, no duplicates.